File: #scripting_languages/javascript/data_structures/disjoint_sets/disjoint_sets.js
/** 
 * Disjoint sets
 * https://www.youtube.com/watch?v=gcmjC-OcWpI&index=14&list=PLlipSLnrfrUlclWAcvmyxcn6R7tzwALhM
 */
function DisjointSet(size){
    var self = this;

    self.size = size;
    self.array = [];
    self.init = init;
    self.union = union;
    self.find = find;
    self.print = print;
    self.init();

    function init(){
        for(var i = 0; i < this.size; i++){
            self.array[i] = -1;
        }
    }

    function union(x, y){
        var xP = find(x);
        var yP = find(y);

        if(self.array[xP] < self.array[yP]){
            self.array[y] = xP;
            self.array[xP]--;
        }else if(self.array[xP] > self.array[yP]){
            self.array[x] = yP;
            self.array[yP]--;
        }else if(self.array[xP] == self.array[yP]){
            self.array[x] = yP;
            self.array[yP]--;
        }
    }

    function find(x){
        if(self.array[x] < 0){
            return x;
        }
        // there was a bug whe I did not use the return on the function call...
        return find(self.array[x]); 
    }

    function print(){
        for(var i = 0; i < this.size; i++){
            console.log(i + ' [' + self.array[i] + ']');
        }
    }
}


var set = new DisjointSet(10);

set.union(1,2);
set.union(2,3);
set.union(3,4);
set.union(4,5);


set.find(1); // 2
set.find(2); // 2
set.find(3); // 2
set.find(4); // 2
set.find(5); // 2

set.find(6); // 6
set.find(7); // 7
set.find(8); // 8
set.find(9); // 9

File: #scripting_languages/ruby/data_structures/disjoint sets/disjoint_sets.rb
class DisjointSet
    def initialize(size)
        @size = size
        @array = []
        @size.times do |index|
            @array[index] = -1
        end
    end

    def union(x, y)
        xP = find(x)
        yP = find(y)

        if @array[xP] < @array[yP]
            @array[y] = xP
            @array[xP] -= 1
        elsif @array[xP] > @array[yP] 
            @array[x] = yP
            @array[yP] -= 1
        elsif @array[xP] == @array[yP]
            @array[x] = yP
            @array[yP] -= 1
        end   
    end

    def find(x)
        if @array[x] < 0
            return x
        end

        # there was a bug whe I did not use the return on the function call...
        return find(@array[x])
    end

    def print
        @size.times do |index|
            puts "#{index} [#{@array[index]}]"
        end
    end
end


set = DisjointSet.new(10)

set.union(1,2)
set.union(2,3)
set.union(3,4)
set.union(4,5)

set.find(1) # 2
set.find(2) # 2
set.find(3) # 2
set.find(4) # 2
set.find(5) # 2

set.find(6) # 6
set.find(7) # 7
set.find(8) # 8
set.find(9) # 9


set.print
# OUTPUT:
#
# 0 [-1]
# 1 [2]
# 2 [-5]
# 3 [2]
# 4 [2]
# 5 [2]
# 6 [-1]
# 7 [-1]
# 8 [-1]
# 9 [-1]
